packages/slate/src/models/node.js

```diff
@@ -1053,17 +1053,12 @@ class Node {
     // If the range is collapsed, check the character before the start.
     if (range.isCollapsed) {
       const text = this.getDescendant(startKey)
-      const char = text.characters.get(range.startOffset - 1)
+      const char = text.characters.get(startOffset - 1)
       return char.marks.toArray()
     }
 
-    // Otherwise, get a set of the marks for each character in the range.
-    return this
-      .getCharactersAtRange(range)
-      .reduce((memo, char) => {
-        char.marks.toArray().forEach(c => memo.push(c))
-        return memo
-      }, [])
+    // Otherwise, return the marks of the first character in the range.
+    return this.getCharactersAtRange(range).first().marks.toArray()
```

Comment:
@blakeembrey this change here sounds to me like it changes how the `getMarksAtRange` method works, so it will affect any place that is using it, [including `value.marks`](https://github.com/ianstormtaylor/slate/blob/master/packages/slate/src/models/value.js#L521-L525) (which is often used for toolbar active states, etc.)

Am I misunderstanding that? It seems like this is fixing the "expanded inserting" case by changing the behavior for all cases?

Reply:
So should I add a new method for this behaviour? This seems like the ideal behaviour as it matched other rich text editors.

Reply:
I think we might not need a new method. Instead it sounds like the `insertTextAtRange` should be changed to not use this method without first changing the range to only include one character? Or something similar. If that ends up requiring a method to control complexity it might be warranted. I think the fix is more related to the text insertion behavior than the mark reading behavior.

Reply:
IIRC, I tried this, but the `.marks` getter uses this method and that's re-used everywhere - should I create a new getter and changed the `.marks` getter behaviour?

Reply:
Now I recall more, it's an issue with pasting also because it uses `document.getMarksAtRange(selection.collapseToStart())` which results in the wrong marks. When you paste with a selection, the selection is being collapsed which results in using the `isCollapsed` logic to select the previous character. In Pages and Word, however, the behaviour is to mirror the selection of the first character - this would need to be a new method just for handling the complexity and I'd need to move pasting, character insertion and other methods over to it. Happy to see how that looks now though, updating the PR ASAP.